lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -1049,52 +1139,76 @@ def struct_from_ruby(hash)
-      if @service_name == CLOUDFUNCTIONS_SERVICE && @cloudfunctions_log_match
+      if resource_type == CLOUDFUNCTIONS_RESOURCE_TYPE &&
+         @cloudfunctions_log_match
         entry.text_payload = convert_to_utf8(
           @cloudfunctions_log_match['text'])
-      elsif @service_name == CLOUDFUNCTIONS_SERVICE && record.key?('log')
+      elsif resource_type == CLOUDFUNCTIONS_RESOURCE_TYPE && record.key?('log')
         entry.text_payload = convert_to_utf8(record['log'])
       elsif is_json
-        entry.struct_payload = struct_from_ruby(record)
-      elsif @service_name == CONTAINER_SERVICE && record.key?('log')
+        entry.json_payload = struct_from_ruby(record)
+      elsif resource_type == CONTAINER_RESOURCE_TYPE && record.key?('log')
         entry.text_payload = convert_to_utf8(record['log'])
       elsif record.size == 1 && record.key?('message')
         entry.text_payload = convert_to_utf8(record['message'])
       else
-        entry.struct_payload = struct_from_ruby(record)
+        entry.json_payload = struct_from_ruby(record)
       end
     end
 
-    def log_name(tag, common_labels)
-      if @service_name == CLOUDFUNCTIONS_SERVICE
+    def log_name(tag, resource_type, resource_labels)
+      if resource_type == CLOUDFUNCTIONS_RESOURCE_TYPE
         return 'cloud-functions'
       elsif @running_on_managed_vm
         # Add a prefix to Managed VM logs to prevent namespace collisions.
-        return "#{APPENGINE_SERVICE}/#{tag}"
-      elsif @service_name == CONTAINER_SERVICE
+        return "#{APPENGINE_SERVICE}%2F#{tag}"
+      elsif resource_type == CONTAINER_RESOURCE_TYPE
         # For Kubernetes logs, use just the container name as the log name
         # if we have it.
-        container_name_key = "#{CONTAINER_SERVICE}/container_name"
-        if common_labels && common_labels.key?(container_name_key)
-          return common_labels[container_name_key]
+        if resource_labels && resource_labels.key?('container_name')
+          return resource_labels['container_name']
         end
       end
       tag
     end
 
+    # Some services set labels (via configuring 'labels' or 'label_map') which
+    # are now MonitoredResource labels in v2.
+    # For these services, remove resource labels from 'labels' and return a
+    # Hash of labels to be merged into the MonitoredResource labels.
+    # Otherwise, return an empty hash and leave 'labels' unmodified.
+    def extract_resource_labels(resource_type, labels)
+      extracted_labels = {}
+      if resource_type == DATAFLOW_RESOURCE_TYPE
+        label_prefix = DATAFLOW_SERVICE
+        labels_to_extract = %w(region job_name job_id step_id)
+      elsif resource_type == ML_RESOURCE_TYPE
+        label_prefix = ML_SERVICE
+        labels_to_extract = %w(job_id task_name)
+      else
+        return extracted_labels
+      end
+
+      labels_to_extract.each do |label|
+        extracted_labels[label] = labels.delete("#{label_prefix}/#{label}") if
```

Comment:
`labels` would never be `nil` in our case, since we have:

```
  case @platform
      when Platform::GCE
        common_labels["#{COMPUTE_SERVICE}/resource_name"] = @vm_name
      when Platform::EC2
        common_labels["#{EC2_SERVICE}/resource_name"] = @vm_name
      when Platform::OTHER
        common_labels["#{COMPUTE_SERVICE}/resource_name"] = @vm_name
      end
```

But it does not hurt to add an extra check to return `{}` if `labels` happens to be `nil` or not a hash. Just added.